From 67f121cc4c3865aa3a93cf563caa1d9da3c92695 Mon Sep 17 00:00:00 2001
From: David Redondo <qt@david-redondo.de>
Date: Mon, 21 Oct 2024 09:22:30 +0200
Subject: [PATCH] client: Run waylandscanner with private-code

This is the recommended mode, and failing to use private-code
makes libraries expose generated symbols like wl_shm_interface,
which may conflict with symbols from libwayland leading to
unexpected behavior or aborts.

Change-Id: Ifcb701a8670b339d24b13f2812c10493a64ee060
Reviewed-by: Eskil Abrahamsen Blomfeldt <eskil.abrahamsen-blomfeldt@qt.io>
---
 examples/wayland/custom-extension/cpp-client/CMakeLists.txt  | 1 +
 examples/wayland/custom-shell/client-plugin/CMakeLists.txt   | 1 +
 src/client/CMakeLists.txt                                    | 1 +
 src/imports/texture-sharing/CMakeLists.txt                   | 2 ++
 .../hardwareintegration/client/brcm-egl/CMakeLists.txt       | 1 +
 .../hardwareintegration/client/dmabuf-server/CMakeLists.txt  | 4 +++-
 .../hardwareintegration/client/drm-egl-server/CMakeLists.txt | 4 +++-
 .../client/libhybris-egl-server/CMakeLists.txt               | 1 +
 .../client/shm-emulation-server/CMakeLists.txt               | 4 +++-
 .../hardwareintegration/client/vulkan-server/CMakeLists.txt  | 5 ++++-
 src/plugins/platforms/qwayland-brcm-egl/CMakeLists.txt       | 1 +
 .../shellintegration/fullscreen-shell-v1/CMakeLists.txt      | 1 +
 src/plugins/shellintegration/ivi-shell/CMakeLists.txt        | 1 +
 src/plugins/shellintegration/qt-shell/CMakeLists.txt         | 1 +
 src/plugins/shellintegration/wl-shell/CMakeLists.txt         | 1 +
 src/plugins/shellintegration/xdg-shell/CMakeLists.txt        | 1 +
 tests/auto/client/clientextension/CMakeLists.txt             | 2 ++
 tests/auto/compositor/compositor/CMakeLists.txt              | 1 +
 tests/manual/server-buffer/cpp-client/CMakeLists.txt         | 1 +
 tests/manual/texture-sharing/cpp-client/CMakeLists.txt       | 1 +
 20 files changed, 31 insertions(+), 4 deletions(-)

diff --git a/examples/wayland/custom-extension/cpp-client/CMakeLists.txt b/examples/wayland/custom-extension/cpp-client/CMakeLists.txt
index 1687ade05..8754719cc 100644
--- a/examples/wayland/custom-extension/cpp-client/CMakeLists.txt
+++ b/examples/wayland/custom-extension/cpp-client/CMakeLists.txt
@@ -20,6 +20,7 @@ qt_add_executable(custom-extension-cpp-client
 )
 
 qt6_generate_wayland_protocol_client_sources(custom-extension-cpp-client
+    PRIVATE_CODE
     FILES
         ${CMAKE_CURRENT_SOURCE_DIR}/../protocol/custom.xml
 )
diff --git a/examples/wayland/custom-shell/client-plugin/CMakeLists.txt b/examples/wayland/custom-shell/client-plugin/CMakeLists.txt
index 39569b80f..71973bf2e 100644
--- a/examples/wayland/custom-shell/client-plugin/CMakeLists.txt
+++ b/examples/wayland/custom-shell/client-plugin/CMakeLists.txt
@@ -22,6 +22,7 @@ target_sources(exampleshellplugin PRIVATE
 )
 
 qt6_generate_wayland_protocol_client_sources(exampleshellplugin
+    PRIVATE_CODE
     FILES
         ${CMAKE_CURRENT_SOURCE_DIR}/../protocol/example-shell.xml
 )
diff --git a/src/client/CMakeLists.txt b/src/client/CMakeLists.txt
index b90ec57c4..eca512877 100644
--- a/src/client/CMakeLists.txt
+++ b/src/client/CMakeLists.txt
@@ -94,6 +94,7 @@ qt_internal_add_module(WaylandClient
 )
 
 qt6_generate_wayland_protocol_client_sources(WaylandClient
+    PRIVATE_CODE
     FILES
         ${CMAKE_CURRENT_SOURCE_DIR}/../3rdparty/protocol/cursor-shape/cursor-shape-v1.xml
         ${CMAKE_CURRENT_SOURCE_DIR}/../3rdparty/protocol/pointer-gestures/pointer-gestures-unstable-v1.xml
diff --git a/src/imports/texture-sharing/CMakeLists.txt b/src/imports/texture-sharing/CMakeLists.txt
index 9b5eb0840..e60f3c2bc 100644
--- a/src/imports/texture-sharing/CMakeLists.txt
+++ b/src/imports/texture-sharing/CMakeLists.txt
@@ -37,8 +37,10 @@ qt_internal_add_qml_module(WaylandTextureSharing
 )
 
 qt6_generate_wayland_protocol_client_sources(WaylandTextureSharing
+    PRIVATE_CODE
     FILES
         ${CMAKE_CURRENT_SOURCE_DIR}/../../extensions/qt-texture-sharing-unstable-v1.xml
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../extensions/server-buffer-extension.xml
 )
 
 qt_internal_add_autogen_sync_header_dependencies(WaylandTextureSharing WaylandCompositor)
diff --git a/src/plugins/hardwareintegration/client/brcm-egl/CMakeLists.txt b/src/plugins/hardwareintegration/client/brcm-egl/CMakeLists.txt
index 6a376ddad..f706632c0 100644
--- a/src/plugins/hardwareintegration/client/brcm-egl/CMakeLists.txt
+++ b/src/plugins/hardwareintegration/client/brcm-egl/CMakeLists.txt
@@ -28,6 +28,7 @@ qt_internal_add_plugin(QWaylandBrcmEglClientBufferPlugin
 )
 
 qt6_generate_wayland_protocol_client_sources(QWaylandBrcmEglClientBufferPlugin
+    PRIVATE_CODE
     FILES
         ${CMAKE_CURRENT_SOURCE_DIR}/../../../../hardwareintegration/client/brcm-egl/../../../extensions/brcm.xml
 )
diff --git a/src/plugins/hardwareintegration/client/dmabuf-server/CMakeLists.txt b/src/plugins/hardwareintegration/client/dmabuf-server/CMakeLists.txt
index bc6c29ecf..7329aebc3 100644
--- a/src/plugins/hardwareintegration/client/dmabuf-server/CMakeLists.txt
+++ b/src/plugins/hardwareintegration/client/dmabuf-server/CMakeLists.txt
@@ -26,8 +26,10 @@ qt_internal_add_plugin(DmaBufServerBufferPlugin
 )
 
 qt6_generate_wayland_protocol_client_sources(DmaBufServerBufferPlugin
+    PRIVATE_CODE
     FILES
-        ${CMAKE_CURRENT_SOURCE_DIR}/../../../../hardwareintegration/client/dmabuf-server/../../../extensions/qt-dmabuf-server-buffer.xml
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../../extensions/qt-dmabuf-server-buffer.xml
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../../extensions/server-buffer-extension.xml
 )
 
 #### Keys ignored in scope 1:.:.:dmabuf-server.pro:<TRUE>:
diff --git a/src/plugins/hardwareintegration/client/drm-egl-server/CMakeLists.txt b/src/plugins/hardwareintegration/client/drm-egl-server/CMakeLists.txt
index 1b40b7fc0..535b1024f 100644
--- a/src/plugins/hardwareintegration/client/drm-egl-server/CMakeLists.txt
+++ b/src/plugins/hardwareintegration/client/drm-egl-server/CMakeLists.txt
@@ -26,8 +26,10 @@ qt_internal_add_plugin(DrmEglServerBufferPlugin
 )
 
 qt6_generate_wayland_protocol_client_sources(DrmEglServerBufferPlugin
+    PRIVATE_CODE
     FILES
-        ${CMAKE_CURRENT_SOURCE_DIR}/../../../../hardwareintegration/client/drm-egl-server/../../../extensions/drm-egl-server-buffer.xml
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../../extensions/drm-egl-server-buffer.xml
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../../extensions/server-buffer-extension.xml
 )
 
 #### Keys ignored in scope 1:.:.:drm-egl-server.pro:<TRUE>:
diff --git a/src/plugins/hardwareintegration/client/libhybris-egl-server/CMakeLists.txt b/src/plugins/hardwareintegration/client/libhybris-egl-server/CMakeLists.txt
index 5167cef49..6b7cd897e 100644
--- a/src/plugins/hardwareintegration/client/libhybris-egl-server/CMakeLists.txt
+++ b/src/plugins/hardwareintegration/client/libhybris-egl-server/CMakeLists.txt
@@ -25,6 +25,7 @@ qt_internal_add_plugin(LibHybrisEglServerBufferPlugin
 )
 
 qt6_generate_wayland_protocol_client_sources(LibHybrisEglServerBufferPlugin
+    PRIVATE_CODE
     FILES
         ${CMAKE_CURRENT_SOURCE_DIR}/../../../../hardwareintegration/client/libhybris-egl-server/../../../extensions/libhybris-egl-server-buffer.xml
 )
diff --git a/src/plugins/hardwareintegration/client/shm-emulation-server/CMakeLists.txt b/src/plugins/hardwareintegration/client/shm-emulation-server/CMakeLists.txt
index 351f7bc81..b979086d2 100644
--- a/src/plugins/hardwareintegration/client/shm-emulation-server/CMakeLists.txt
+++ b/src/plugins/hardwareintegration/client/shm-emulation-server/CMakeLists.txt
@@ -25,8 +25,10 @@ qt_internal_add_plugin(ShmServerBufferPlugin
 )
 
 qt6_generate_wayland_protocol_client_sources(ShmServerBufferPlugin
+    PRIVATE_CODE
     FILES
-        ${CMAKE_CURRENT_SOURCE_DIR}/../../../../hardwareintegration/client/shm-emulation-server/../../../extensions/shm-emulation-server-buffer.xml
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../../extensions/server-buffer-extension.xml
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../../extensions/shm-emulation-server-buffer.xml
 )
 
 #### Keys ignored in scope 1:.:.:shm-emulation-server.pro:<TRUE>:
diff --git a/src/plugins/hardwareintegration/client/vulkan-server/CMakeLists.txt b/src/plugins/hardwareintegration/client/vulkan-server/CMakeLists.txt
index 72f527210..f0aeabf29 100644
--- a/src/plugins/hardwareintegration/client/vulkan-server/CMakeLists.txt
+++ b/src/plugins/hardwareintegration/client/vulkan-server/CMakeLists.txt
@@ -25,8 +25,11 @@ qt_internal_add_plugin(VulkanServerBufferPlugin
 )
 
 qt6_generate_wayland_protocol_client_sources(VulkanServerBufferPlugin
+    PRIVATE_CODE
     FILES
-        ${CMAKE_CURRENT_SOURCE_DIR}/../../../../hardwareintegration/client/vulkan-server/../../../extensions/qt-vulkan-server-buffer-unstable-v1.xml
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../../extensions/qt-vulkan-server-buffer-unstable-v1.xml
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../../extensions/server-buffer-extension.xml
+
 )
 
 #### Keys ignored in scope 1:.:.:vulkan-server.pro:<TRUE>:
diff --git a/src/plugins/platforms/qwayland-brcm-egl/CMakeLists.txt b/src/plugins/platforms/qwayland-brcm-egl/CMakeLists.txt
index 6289fc1ff..ac6cef722 100644
--- a/src/plugins/platforms/qwayland-brcm-egl/CMakeLists.txt
+++ b/src/plugins/platforms/qwayland-brcm-egl/CMakeLists.txt
@@ -28,6 +28,7 @@ qt_internal_add_plugin(QWaylandBrcmEglPlatformIntegrationPlugin
 )
 
 qt6_generate_wayland_protocol_client_sources(QWaylandBrcmEglPlatformIntegrationPlugin
+    PRIVATE_CODE
     FILES
         ${CMAKE_CURRENT_SOURCE_DIR}/../../../hardwareintegration/client/brcm-egl/../../../extensions/brcm.xml
 )
diff --git a/src/plugins/shellintegration/fullscreen-shell-v1/CMakeLists.txt b/src/plugins/shellintegration/fullscreen-shell-v1/CMakeLists.txt
index 693dbde5f..9f16e82d9 100644
--- a/src/plugins/shellintegration/fullscreen-shell-v1/CMakeLists.txt
+++ b/src/plugins/shellintegration/fullscreen-shell-v1/CMakeLists.txt
@@ -26,6 +26,7 @@ qt_internal_add_plugin(QWaylandFullScreenShellV1IntegrationPlugin
 )
 
 qt6_generate_wayland_protocol_client_sources(QWaylandFullScreenShellV1IntegrationPlugin
+    PRIVATE_CODE
     FILES
         ${CMAKE_CURRENT_SOURCE_DIR}/../../../3rdparty/protocol/fullscreen-shell/fullscreen-shell-unstable-v1.xml
 )
diff --git a/src/plugins/shellintegration/ivi-shell/CMakeLists.txt b/src/plugins/shellintegration/ivi-shell/CMakeLists.txt
index cfdd0e660..0e661c5e9 100644
--- a/src/plugins/shellintegration/ivi-shell/CMakeLists.txt
+++ b/src/plugins/shellintegration/ivi-shell/CMakeLists.txt
@@ -26,6 +26,7 @@ qt_internal_add_plugin(QWaylandIviShellIntegrationPlugin
 )
 
 qt6_generate_wayland_protocol_client_sources(QWaylandIviShellIntegrationPlugin
+    PRIVATE_CODE
     FILES
         ${CMAKE_CURRENT_SOURCE_DIR}/../../../3rdparty/protocol/ivi/ivi-application.xml
         ${CMAKE_CURRENT_SOURCE_DIR}/../../../3rdparty/protocol/ivi/ivi-controller.xml
diff --git a/src/plugins/shellintegration/qt-shell/CMakeLists.txt b/src/plugins/shellintegration/qt-shell/CMakeLists.txt
index 9de110f69..050231733 100644
--- a/src/plugins/shellintegration/qt-shell/CMakeLists.txt
+++ b/src/plugins/shellintegration/qt-shell/CMakeLists.txt
@@ -22,6 +22,7 @@ qt_internal_add_plugin(QWaylandQtShellIntegrationPlugin
 )
 
 qt6_generate_wayland_protocol_client_sources(QWaylandQtShellIntegrationPlugin
+    PRIVATE_CODE
     FILES
         ${CMAKE_CURRENT_SOURCE_DIR}/../../../extensions/qt-shell-unstable-v1.xml
 )
diff --git a/src/plugins/shellintegration/wl-shell/CMakeLists.txt b/src/plugins/shellintegration/wl-shell/CMakeLists.txt
index b2871c92b..6ed139295 100644
--- a/src/plugins/shellintegration/wl-shell/CMakeLists.txt
+++ b/src/plugins/shellintegration/wl-shell/CMakeLists.txt
@@ -28,6 +28,7 @@ qt_internal_add_module(WlShellIntegrationPrivate
 qt6_generate_wayland_protocol_client_sources(WlShellIntegrationPrivate
     # this is just a helper module and still part of the QtWaylandClient build
     __QT_INTERNAL_WAYLAND_INCLUDE_DIR QtWaylandClient/private
+    PRIVATE_CODE
     FILES
         ${CMAKE_CURRENT_SOURCE_DIR}/../../../3rdparty/protocol/wayland/wayland.xml
 )
diff --git a/src/plugins/shellintegration/xdg-shell/CMakeLists.txt b/src/plugins/shellintegration/xdg-shell/CMakeLists.txt
index 74e925f73..529ea1f6b 100644
--- a/src/plugins/shellintegration/xdg-shell/CMakeLists.txt
+++ b/src/plugins/shellintegration/xdg-shell/CMakeLists.txt
@@ -34,6 +34,7 @@ qt_internal_add_plugin(QWaylandXdgShellIntegrationPlugin
 )
 
 qt6_generate_wayland_protocol_client_sources(QWaylandXdgShellIntegrationPlugin
+    PRIVATE_CODE
     FILES
         ${CMAKE_CURRENT_SOURCE_DIR}/../../../3rdparty/protocol/xdg-decoration/xdg-decoration-unstable-v1.xml
         ${CMAKE_CURRENT_SOURCE_DIR}/../../../3rdparty/protocol/xdg-shell/xdg-shell.xml
diff --git a/tests/auto/client/clientextension/CMakeLists.txt b/tests/auto/client/clientextension/CMakeLists.txt
index 4997b7d77..063f3f474 100644
--- a/tests/auto/client/clientextension/CMakeLists.txt
+++ b/tests/auto/client/clientextension/CMakeLists.txt
@@ -9,9 +9,11 @@ qt_internal_add_test(tst_clientextension
 )
 
 qt6_generate_wayland_protocol_client_sources(tst_clientextension
+    PRIVATE_CODE
     FILES ${CMAKE_CURRENT_SOURCE_DIR}/test.xml
 )
 
 qt6_generate_wayland_protocol_server_sources(tst_clientextension
+    PRIVATE_CODE
     FILES ${CMAKE_CURRENT_SOURCE_DIR}/test.xml
 )
diff --git a/tests/auto/compositor/compositor/CMakeLists.txt b/tests/auto/compositor/compositor/CMakeLists.txt
index 701e8731d..2ea395be7 100644
--- a/tests/auto/compositor/compositor/CMakeLists.txt
+++ b/tests/auto/compositor/compositor/CMakeLists.txt
@@ -29,6 +29,7 @@ qt_internal_add_test(tst_compositor
 )
 
 qt6_generate_wayland_protocol_client_sources(tst_compositor
+    PRIVATE_CODE
     FILES
         ${CMAKE_CURRENT_SOURCE_DIR}/../../../../src/3rdparty/protocol/idle-inhibit/idle-inhibit-unstable-v1.xml
         ${CMAKE_CURRENT_SOURCE_DIR}/../../../../src/3rdparty/protocol/ivi/ivi-application.xml
diff --git a/tests/manual/server-buffer/cpp-client/CMakeLists.txt b/tests/manual/server-buffer/cpp-client/CMakeLists.txt
index 01c0df44f..0b58b1b07 100644
--- a/tests/manual/server-buffer/cpp-client/CMakeLists.txt
+++ b/tests/manual/server-buffer/cpp-client/CMakeLists.txt
@@ -20,6 +20,7 @@ qt_add_executable(server-buffer-cpp-client
 )
 
 qt6_generate_wayland_protocol_client_sources(server-buffer-cpp-client
+    PRIVATE_CODE
     FILES
         ${CMAKE_CURRENT_SOURCE_DIR}/../share-buffer.xml
 )
diff --git a/tests/manual/texture-sharing/cpp-client/CMakeLists.txt b/tests/manual/texture-sharing/cpp-client/CMakeLists.txt
index 1059e6e5e..1a9c1b522 100644
--- a/tests/manual/texture-sharing/cpp-client/CMakeLists.txt
+++ b/tests/manual/texture-sharing/cpp-client/CMakeLists.txt
@@ -23,6 +23,7 @@ qt_internal_add_manual_test(cpp-client
 )
 
 qt6_generate_wayland_protocol_client_sources(cpp-client
+    PRIVATE_CODE
     FILES
         ${CMAKE_CURRENT_SOURCE_DIR}/../../../../src/extensions/qt-texture-sharing-unstable-v1.xml
 )
